File: cpp/src/ch05/monte_carlo.cc
#include "monte_carlo.h"
#include "random_action.h"

using std::cout;
using std::endl;

double playout(State *state)
{
    if (state->is_done())
        return state->teban_score();

    state->advance(random_action(*state));
    // win score always returns score of player A
    // so no need to do 1 - playout(state)
    return 1 - playout(state);
}

int primitive_monte_carlo_action(const State &state,
                                 int playout_number)
{
    auto legal_actions = state.legal_actions();
    auto values = std::vector<double>(legal_actions.size());
    auto cnts = std::vector<double>(legal_actions.size());
    for (int cnt = 0; cnt < playout_number; cnt++)
    {
        int index = cnt % legal_actions.size();
        State next_state = state;
        next_state.advance(legal_actions[index]);
        values[index] += 1 - playout(&next_state);
        ++cnts[index];
    }
    int best_action_index = -1;
    double best_score = -INF;
    for (int index = 0; index < legal_actions.size(); index++)
    {
        double value_mean = values[index] / cnts[index];
        if (value_mean > best_score)
        {
            best_score = value_mean;
            best_action_index = index;
        }
    }

    // print average score of each action for debugging
    /*
    auto average_score = std::vector<double>(legal_actions.size());
    cout << "average" << endl;
    for (int i = 0; i < legal_actions.size(); i++)
    {
        average_score[i] = values[i] / cnts[i];
        cout << i << " " << legal_actions[i] << " " << average_score[i] << endl;
    }

    cout << "best action " << best_action_index << " "
         << legal_actions[best_action_index] << endl
         << endl;
    */

    return legal_actions[best_action_index];
}

File: rust/src/ch05/monte_carlo.rs
use std::sync::Arc;

use crate::base::alternate::{ActionFunc, AlternateState};
use crate::ch05::random_action;

// implements playout, because playout needs to own the mutable state
// make it a struct not a function
pub struct Playout<T: AlternateState> {
    state: T,
}

impl<T: AlternateState> Playout<T> {
    pub fn new(state: &T) -> Self {
        Playout {
            state: state.clone(),
        }
    }

    pub fn playout(&mut self) -> f32 {
        if self.state.is_done() {
            return self.state.white_score();
        }

        let action = random_action::random_action_arc()(&self.state);
        self.state.advance(action);
        1.0 - self.playout()
    }
}

fn monte_carlo_action<T: AlternateState>(
    initial_state: &T,
    num_playout: usize,
) -> usize {
    let legal_actions = initial_state.legal_actions();
    let mut values = vec![0.0; legal_actions.len()];
    let mut counts: Vec<usize> = vec![0; legal_actions.len()];
    for count in 0..num_playout {
        let index = count % legal_actions.len();
        let mut state = initial_state.clone();
        state.advance(legal_actions[index]);
        // 2回　clone することになるので無駄っぽいけど、
        // mutate するバグを以前踏んだのでこのまま
        values[index] += 1.0 - Playout::new(&state).playout();
        counts[index] += 1;
    }

    // Calculate the action-score pairs
    let action_scores: Vec<(usize, f32)> = legal_actions
        .iter()
        .zip(values.iter().zip(counts.iter()))
        .map(|(&action, (&value, &count))| (action, value / count as f32))
        .collect();

    // Find the best action based on the scores
    action_scores
        .iter()
        .max_by(|(_, score1), (_, score2)| score1.partial_cmp(&score2).unwrap())
        .unwrap()
        .0
        .to_owned()
}

pub fn monte_carlo_action_arc<T: AlternateState>(
    num_playout: usize,
) -> ActionFunc<T> {
    Arc::new(move |state| -> usize { monte_carlo_action(state, num_playout) })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::base::alternate::MazeParams;
    use crate::ch05::maze_state;

    fn setup() -> maze_state::AlternateMazeState {
        let params = MazeParams {
            height: 3,
            width: 3,
            end_turn: 4,
        };
        maze_state::AlternateMazeState::new(0, params)
    }

    #[test]
    fn test_monte_carlo() {
        let state = setup();
        let actual = monte_carlo_action(&state, 100);
        let expected = 0;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_monte_carlo_arc() {
        let state = setup();
        let actual = monte_carlo_action_arc(100)(&state);
        let expected = 0;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_playout() {
        let state = setup();
        let mut playout = Playout::new(&state);
        let score = playout.playout();
        // check that state isn't mutated
        assert_eq!(state.turn, 0);
        // result is random
        assert!(score <= 1.0);
    }
}
